### PR TITLE
JAMES-3355: Jmap set delete

### PR DESCRIPTION
**Pull request by nvduc91**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala [hunk 1]
```diff
@@ -28,7 +28,7 @@ import org.apache.james.jmap.model.CapabilityIdentifier.CapabilityIdentifier
 import org.apache.james.jmap.model.Invocation.{Arguments, MethodName}
 import org.apache.james.jmap.model.{Invocation, State}
 import org.apache.james.mailbox.exception.{InsufficientRightsException, MailboxExistsException, MailboxNameException, MailboxNotFoundException}
-import org.apache.james.mailbox.model.{MailboxId, MailboxPath}
+import org.apache.james.mailbox.model._
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:31]: Wildcard

**Comment by nvduc91** [comment]
add missing import while revert wildcard.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala [hunk 2]
```diff
@@ -38,16 +38,13 @@ import reactor.core.scheduler.Schedulers
 
 import scala.collection.immutable
 
+case class MailboxHasMail(mailboxId: MailboxId) extends Exception
+case class MailboxHasChild(mailboxId: MailboxId) extends Exception
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:42]: May be we should rename these to MailboxHasMailException & MailboxHasChildException

**Comment by nvduc91** [comment]
rebase https://github.com/linagora/james-project/pull/3656 (delete `toSet` method)

**Comment by nvduc91** [comment]
`
[15bdf96e5383661ba065c63edb2f956330c9c07d] docker: Error response from daemon: No such container: git-publish-jmap-set-delete-9758abe2-fa70-4f64-8a5f-af32d05e98f6.
`
Test green.
test this please

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 3]
```diff
@@ -669,4 +669,204 @@ trait MailboxSetMethodContract {
          |    "c1"]]
          |}""".stripMargin)
   }
+
+  @Test
+  def deleteShouldSuccessWhenMailboxExists(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:680]:
```suggestion
  def deleteShouldSucceedWhenMailboxExists(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 4]
```diff
@@ -869,4 +872,57 @@ trait MailboxSetMethodContract {
          |    ]
          |}""".stripMargin)
   }
+
+  @Test
+  def deleteShouldFailWhenMailboxDoesNotExists(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:881]:
```suggestion
  def deleteShouldFailWhenMailboxDoesNotExist(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala [hunk 5]
```diff
@@ -1043,4 +1043,60 @@ trait MailboxSetMethodContract {
          |    "c1"]]
          |}""".stripMargin)
   }
+
+  @Test
+  def deleteShouldFailWhenNotEnoughRight(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala:1048]:
```suggestion
  def deleteShouldFailWhenNotEnoughRights(server: GuiceJamesServer): Unit = {
```

**Comment by nvduc91** [comment]
Im handling conflict.

**Comment by nvduc91** [comment]
rebase master

**Comment by chibenwa** [comment]
```
[ERROR]   RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultEndpointShouldRestoreImapDeletedMailbox:252 » ConditionTimeout
```

File: server/protocols/jmap-rfc-8621-integration-tests/memory-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/memory/MemoryMailboxSetMethodTest.java [hunk 6]
```diff
@@ -21,14 +21,20 @@
 
 import static org.apache.james.MemoryJamesServerMain.IN_MEMORY_SERVER_AGGREGATE_MODULE;
 
+<<<<<<< 71a6e3bfa1f9baca3c6ea9ceab17f6ac21dce2e6
 import java.util.concurrent.ThreadLocalRandom;
 
+=======
+>>>>>>> [NO REVIEW] add logger for internal error in jmapApiRoutes
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/memory-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/memory/MemoryMailboxSetMethodTest.java]: Oups

**Comment by chibenwa** [comment]
`Caused by: java.net.BindException: Address already in use`

test this please

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala [hunk 7]
```diff
@@ -34,11 +34,12 @@ case class MailboxSetRequest(accountId: AccountId,
                              ifInState: Option[State],
                              create: Option[Map[MailboxCreationId, JsObject]],
                              update: Option[Map[MailboxId, MailboxPatchObject]],
-                             destroy: Option[Seq[MailboxId]],
+                             destroy: Option[Seq[UnparsedMailboxId]],
                              onDestroyRemoveEmails: Option[RemoveEmailsOnDestroy])
 
 object MailboxSetRequest {
   type MailboxCreationId = String Refined NonEmpty
+  type UnparsedMailboxId = String Refined NonEmpty
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala:42]: As `MailboxCreationId` and `UnparsedMailboxId` are exactly the same type, wouldn't it be better to just keep one generic type here?

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala:42]: It is two different usage that happens to share the same constraints yet it is two distinct things.

MailboxCreationId is any non empty string specified by the client

UnparsedMailboxId is # + MailboxCreationId OR a "real" mailboxId

**Comment by chibenwa** [comment]
deployment test

test this please

**Comment by chibenwa** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project james-server-webadmin-data: There was a timeout or other error in the fork -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
```
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 11:59:12.868 [ERROR] s.h.m.GeneralError - 
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] java.lang.RuntimeException: null
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.apache.james.webadmin.service.UserService.upsert(UserService.java:77)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.apache.james.webadmin.service.UserService.upsertUser(UserService.java:65)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.apache.james.webadmin.routes.UserRoutes.upsertUser(UserRoutes.java:225)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.RouteImpl$1.handle(RouteImpl.java:72)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.http.matching.Routes.execute(Routes.java:61)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:134)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.embeddedserver.jetty.JettyHandler.doHandle(JettyHandler.java:50)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1671)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.Server.handle(Server.java:505)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at java.base/java.lang.Thread.run(Thread.java:834)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] Exception in thread "Thread-3683" java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:44395
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.apache.james.webadmin.WebAdminServer.lambda$start$1(WebAdminServer.java:95)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.Service.lambda$init$2(Service.java:582)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at java.base/java.lang.Thread.run(Thread.java:834)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] Caused by: java.io.IOException: Failed to bind to /0.0.0.0:44395
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:346)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:308)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.server.Server.doStart(Server.java:396)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.embeddedserver.jetty.EmbeddedJettyServer.ignite(EmbeddedJettyServer.java:149)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	at spark.Service.lambda$init$2(Service.java:574)
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] 	... 1 more
[cc60295bc1379ecf05c8dd3690c8b019b3c9cc2b] Caused by: java.net.BindException: Address already in use
```

**Comment by chibenwa** [comment]
Merged

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala (hunk 7 above)
Comment (mbaechler) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala:42]: yet the type doesn't encode that knowledge.
You could express that with a sum type (an ADT or an Either)

**Pull request closed without merging**